test/query.test.js

```diff
@@ -0,0 +1,229 @@
+import state from './fixtures/solar-system'
+import Query from '../src/addons/query'
+import { set } from '../src/microcosm'
+
+describe('get', function () {
+
+  it('fetches a value', function () {
+    let query = new Query().get('physics.gravity')
+
+    expect(query.compute(state)).toEqual(1)
+  })
+
+  it('can provide a fallback a value', function () {
+    let query = new Query().get('physics.heat', 10)
+
+    expect(query.compute(state)).toEqual(10)
+  })
+
+  it('can nest arguments', function () {
+    let query = new Query().get('physics.gravity')
+
+    expect(query.compute(state)).toEqual(1)
+  })
+
+  it('gracefully handles recomputing missing keys', function () {
+    let query = new Query().get('physics.foo.bar')
+
+    query.compute(state)
+    query.compute(state)
+  })
+
+  it('aliases to query.select', function () {
+    let query = new Query().select('physics.foo.bar', 10)
+
+    let answer = query.compute(state)
+
+    expect(answer).toEqual(10)
+  })
+
+})
+
+describe('tap', function () {
+
+  it('executes a function', function () {
+    let query = new Query().get('physics.gravity').tap(n => n * 10)
+
+    expect(query.compute(state)).toEqual(10)
+  })
+
+})
+
+describe('filter', function () {
+  it('reduces a data set given a predicate', function () {
+    let query = new Query().get('planets').filter(planet => planet.id === 'pluto').map(i => i.id)
+
+    expect(query.compute(state)).toEqual(['pluto'])
+  })
+
+  it('always returns the same array if the data has not changed', function () {
+    let query = new Query().get('planets').filter(planet => planet.id === 'pluto').map(i => i.id)
+
+    let a = query.compute(state)
+    let b = query.compute(state)
+
+    expect(a).toBe(b)
+  })
+})
+
+describe('at', function () {
+  it('gets an item at an index', function () {
+    let query = new Query().get('planets').at(0).get('id')
+
+    expect(query.compute(state)).toEqual('mercury')
+  })
+
+  it('gets an item from the back of the list', function () {
+    let query = new Query().get('planets').at(-1).get('id')
+
+    expect(query.compute(state)).toEqual('pluto')
+  })
+
+  it('returns undefined if out of the lower boundary', function () {
+    let query = new Query().get('planets').at(-10)
+
+    expect(query.compute(state)).toBeUndefined()
+  })
+
+  it('returns undefined if out of the upper boundary', function () {
+    let query = new Query().get('planets').at(10)
+
+    expect(query.compute(state)).toBeUndefined()
+  })
+})
+
+describe('join', function () {
+  it('filters by a criteria within the dataset', function () {
+    let query = new Query().get('planets').join('id', 'meta.selected').at(0).get('id')
```

Comment:
I hate the name `join` too... Basically, query functions can specify dependencies. The query won't run again if those dependencies haven't changed.

SoI want a way to filter down from another key while declaring that the query depends on the `meta.selected` field. Another option could be to allow `get` to fetch more than one item:

```javascript
q.get('planets', 'meta.selected')
  .tap((planets, selected) => planets.filter(p => p.id === selected)
```

So how do we make that more concise?